File: 2025/NeetCode/Arrays & Hashing/1. Two Sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {

    let map = new Map();

    for(let i = 0; i < nums.length; i++) {
        const compliment = target - nums[i];  
        
        if(map.has(compliment)) {
            return [i, map.get(compliment)];
        } else {
            map.set(nums[i], i);
        }
    }
    return false;
    
};

// Time Complexity = O(n)
// Space Complexity = O(n)

File: 2026/Blind 75 [Python]/Arrays & Hashing/1. Two Sum.py
from typing import List

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        prevMap = {} # val : index

        for i, n in enumerate(nums): # enumerate provides the index along with the number
            diff = target - n
            if diff in prevMap:
                return [prevMap[diff], i]
            prevMap[n] = i # store the index

# Time Complexity: O(N)
# Space Complexity: O(N)